linux/ibus-keyman/test/test_ibus_keyman.py

```diff
@@ -0,0 +1,284 @@
+                    for part in keyparts:
+                        # print(part)
+                        if part in self.known_keys:
+                            mainkey = part
+                            # print("mainkey:", mainkey)
+                        elif part in self.known_modifiers:
+                            mods.append(part)
+                            # print("mods:", mods)
+                        # else:
+                            # print("unknown part %s", part)
+                    logging.info("key is %s with modifiers %s" % (mainkey, mods))
+                    if len(mods) == 0:
+                        localkeyboard.press(self.known_keys[mainkey])
+                        localkeyboard.release(self.known_keys[mainkey])
+                    elif len(mods) == 1:
+                        with localkeyboard.pressed(self.known_modifiers[mods[0]]):
+                            localkeyboard.press(self.known_keys[mainkey])
+                            localkeyboard.release(self.known_keys[mainkey])
+                    elif len(mods) == 2:
+                        with localkeyboard.pressed(self.known_modifiers[mods[0]]):
+                            with localkeyboard.pressed(self.known_modifiers[mods[1]]):
+                                localkeyboard.press(self.known_keys[mainkey])
+                                localkeyboard.release(self.known_keys[mainkey])
+                    elif len(mods) == 3:
+                        with localkeyboard.pressed(self.known_modifiers[mods[0]]):
+                            with localkeyboard.pressed(self.known_modifiers[mods[1]]):
+                                    with localkeyboard.pressed(self.known_modifiers[mods[2]]):
+                                        localkeyboard.press(self.known_keys[mainkey])
+                                        localkeyboard.release(self.known_keys[mainkey])
+                    elif len(mods) == 4:
+                        with localkeyboard.pressed(self.known_modifiers[mods[0]]):
+                            with localkeyboard.pressed(self.known_modifiers[mods[1]]):
+                                with localkeyboard.pressed(self.known_modifiers[mods[2]]):
+                                    with localkeyboard.pressed(self.known_modifiers[mods[3]]):
+                                        localkeyboard.press(self.known_keys[mainkey])
+                                        localkeyboard.release(self.known_keys[mainkey])
+                    else:
+                        logging.warning("too many modifiers %d:%s", len(mods), mods)
+                        localkeyboard.press(self.known_keys[mainkey])
+                        localkeyboard.release(self.known_keys[mainkey])
+                    # Use sleep to slow down keypresses to more natural level
+                    # At full speed ibus_set_surrounding_text events were coming in
+                    # which got text 1 or 2 keypresses behind
+                    time.sleep(0.3)
```

Comment:
This worries me! Will the same problem arise if end users type rapidly? It should be predictable no matter what the input speed is.

Reply:
I don't know. I haven't encountered problems typing myself.

It was a last resort because the only way to know if mouse movement has happened is to check if the context has changed when set_surrounding_context is called. But the gui was getting behind what was sent to it on my laptop which isn't particularly fast.